flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java

```diff
@@ -367,27 +381,39 @@ private void startJobMasterServiceSafely(UUID leaderSessionId) {
                             userCodeClassLoader,
                             initializationTimestamp);
 
+            // Execute listener notification asynchronously in the main thread executor to make sure
+            // this "grant leadership" operation is properly completed before the next operation is
+            // started. With the current implementation DispatcherJob might call closeAsync()
+            // leading to concurrent access under the "lock".
+            FutureUtils.assertNoException(
+                    CompletableFuture.runAsync(
+                            () -> jobManagerStatusListener.onJobManagerStarted(this),
+                            mainThreadExecutor));
```

Comment:
I think there is design problem here. Somehow the `JobManagerRunner` needs to know that some calls have to be executed by the `mainThreadExecutor`. This looks like implementation details which are leaked from the `DispatcherJob` into the `JobManagerRunnerImpl`.